cmd/sourcemaps/upload.go

```diff
@@ -0,0 +1,334 @@
+				files = append(files, SourceMapFile{
+					Path: filePath,
+					Name: routeGroupRemovedPath,
+				})
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(files) == 0 {
+		return nil, fmt.Errorf("no .js.map files found. Please double check that you have generated sourcemaps for your app")
+	}
+
+	return files, nil
+}
+
+func getS3Key(organizationID, version, basePath, fileName string) string {
+	if version == "" {
+		version = "unversioned"
+	}
+
+	if basePath != "" && !strings.HasSuffix(basePath, "/") {
+		basePath = basePath + "/"
+	}
+
+	return fmt.Sprintf("%s/%s/%s%s", organizationID, version, basePath, fileName)
+}
+
+func getSourceMapUploadUrls(apiKey string, paths []string, backendUrl string) ([]string, error) {
+	variables := map[string]interface{}{
+		"api_key": apiKey,
+		"paths":   paths,
+	}
+
+	reqBody, err := json.Marshal(map[string]interface{}{
+		"query":     getSourceMapUrlsQuery,
+		"variables": variables,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", backendUrl, bytes.NewBuffer(reqBody))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var urlsResp SourceMapUrlsResponse
+	if err := json.Unmarshal(body, &urlsResp); err != nil {
+		return nil, err
+	}
+
+	if len(urlsResp.Data.GetSourceMapUploadUrls) == 0 {
+		return nil, fmt.Errorf("unable to generate source map upload urls")
+	}
+
+	return urlsResp.Data.GetSourceMapUploadUrls, nil
+}
+
+func uploadFile(filePath, uploadUrl, name string) error {
+	fileContent, err := os.ReadFile(filePath)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("PUT", uploadUrl, bytes.NewBuffer(fileContent))
+	if err != nil {
+		return err
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("upload failed with status code: %d", resp.StatusCode)
+	}
+
+	fmt.Printf("[LaunchDarkly] Uploaded %s to %s\n", filePath, name)
+	return nil
+}
+
+func initFlags(cmd *cobra.Command) {
+	cmd.Flags().String(apiKeyFlag, "", "The LaunchDarkly Observability API key")
+	_ = cmd.MarkFlagRequired(apiKeyFlag)
+	_ = cmd.Flags().SetAnnotation(apiKeyFlag, "required", []string{"true"})
+	_ = viper.BindPFlag(apiKeyFlag, cmd.Flags().Lookup(apiKeyFlag))
+
+	cmd.Flags().String(appVersionFlag, "", "The current version of your deploy")
+	_ = viper.BindPFlag(appVersionFlag, cmd.Flags().Lookup(appVersionFlag))
+
+	cmd.Flags().String(pathFlag, defaultPath, "Sets the directory of where the sourcemaps are")
+	_ = viper.BindPFlag(pathFlag, cmd.Flags().Lookup(pathFlag))
+
+	cmd.Flags().String(basePathFlag, "", "An optional base path for the uploaded sourcemaps")
```

Comment:
aside

Why does the customer care about this / why would the customer change this? given that they don't have direct access to `defaultBackendUrl`

Reply:
aside This wasn't clear from our old docs either, but the basePath arg is for transforming the destination path on our end. For example:

customer has sourcemaps under `./dist/foo` directory
when they deploy their app, their code runs from `./server/bar`
they should provide `--path ./dist/foo --basePath ./server/bar` so that the source maps are written to the write path that we will map their deployment correctly to.